include/pybind11/cast.h

```diff
@@ -157,7 +157,8 @@ inline void keep_alive_impl(handle nurse, handle patient);
 class type_caster_generic {
 public:
     PYBIND11_NOINLINE type_caster_generic(const std::type_info &type_info)
-     : typeinfo(get_type_info(type_info, false)) { }
+     : typeinfo(get_type_info(type_info, false)),
+       direct_conversions(get_internals().direct_conversions[std::type_index(type_info)]) { }
```

Comment:
Not so excited about the thought of another global hash map lookup here -- these are kind of expensive as was noticed in another context. I'll have to think a bit about this.